[dune] Simple rule to generate Stdlib's documentation.

ejgallego opened the pull request:
Ideally this will be handled by Dune's native library support, but
this could be useful for the likes of #9648.

I am not sure what should be done w.r.t. style files.

rgrinberg commented on doc/stdlib/dune:
```diff
@@ -0,0 +1,17 @@
+   (:footer %{project_root}/doc/common/styles/html/coqremote/footer.html))
+ (action
+  (progn
+   (bash "mkdir -p html")
```
A bit curious why you need bash here. `(run mkdir -p html)` doesn't work?

ejgallego replied:
Good point, thanks I never know if this kind of commands are shell builtins or not.

rgrinberg commented on doc/stdlib/dune:
```diff
@@ -0,0 +1,17 @@
+ (action
+  (progn
+   (bash "mkdir -p html")
+   (bash "%{bin:coqdoc} -q -d html --with-header %{header} --with-footer %{footer} --multi-index --html -g -R %{project_root}/theories Coq -R %{project_root}/plugins Coq $(find %{project_root}/theories %{project_root}/plugins -name *.v)"))))
```
Same here. It should be possible to do this without bash. Just bind the dependencies in theories and plugins to a name. If that doesn't work, I think we should have a bug report for that.

ejgallego replied:
If I bind the dependencies I'll get all the files in the tree, how can I filter just the `.v` ones?

I had to use bash here due to quoting issues when inlining the `find` call.

rgrinberg replied:
I see, so actually, the feature that you need here is a recursive glob. Basically, you want to write:

```
(deps (coq_files (glob_files theories/**/*.v)))
```

I'm not entirely sure if we have a ticket for this, but the feature was mentioned a few times before. I'll look it up.

rgrinberg replied:
Here's the ticket if anyone is curious: https://github.com/ocaml/dune/issues/1866

ejgallego replied:
Added a comment to update in the future, thanks!

SkySkimmer commented on .gitlab-ci.yml:
```diff
@@ -340,7 +340,7 @@ doc:refman:dune:
   variables:
     <<: *dune-ci-template-vars
-    DUNE_TARGET: refman-html
+    DUNE_TARGET: refman-html stdlib-html
```
Any reason to do it in the same job vs adding another job?

ejgallego replied:
Efficiency. One documentation job seems enough IMO. In particular the deploy job needs to download this. That should change if we split the stdlib build off tho.

ejgallego replied:
I think you were right and I've changed this.

SkySkimmer commented on doc/stdlib/dune:
```diff
@@ -0,0 +1,17 @@
+   (source_tree %{project_root}/theories)
+   (source_tree %{project_root}/plugins)
+   (:header %{project_root}/doc/common/styles/html/coqremote/header.html)
+   (:footer %{project_root}/doc/common/styles/html/coqremote/footer.html))
```
Doesn't coqdoc read glob files or some such thing?

ejgallego replied:
Good point, on the current setup they will be there as part of the artifacts but indeed we should add the package dependency on.

When native support comes the right rule will be generated tho.

SkySkimmer commented:
Compcert failed due to https://github.com/AbsInt/CompCert/pull/275#issuecomment-467336477 
Should be fixed now

ejgallego commented:
> Do we merge without the .glob dependencies?

Nope, let me fix that indeed, and rebase on top of #9648

SkySkimmer commented:
I'll let @Zimmi48 as dune secondary assign.

ejgallego commented:
Rebased, but I am not sure how to test the deploy job (@vbgl ?)

vbgl commented:
To test deploy jobs, I’ve manually started a CI pipeline with the appropriate credentials. I can do it for you if you think its necessary.

ejgallego commented:
> To test deploy jobs, I’ve manually started a CI pipeline with the appropriate credentials. I can do it for you if you think its necessary.

Yup but that would overwrite the main Coq repos, right?

I guess I'll deploy to my own repos, and make the clone target a variable then.

vbgl commented:
It will not overwrite, but put you artifact in a separate directory (named as your branch: `dune+coqdoc`).

ejgallego commented:
> It will not overwrite, but put you artifact in a separate directory (named as your branch: `dune+coqdoc`).

Ah Ok, cool then!

ejgallego commented:
I'd appreciate indeed if you can push a build to see whether that works.

vbgl commented:
Pipeline started. Result should appear in a few hours there: https://coq.github.io/doc/dune+coqdoc

ejgallego commented:
> Pipeline started. Result should appear in a few hours there: https://coq.github.io/doc/dune+coqdoc

Muchas gracias @vbgl

Zimmi48 commented on Makefile.dune:
```diff
@@ -4,7 +4,7 @@
 .PHONY: coq coqide coqide-server                      # Package targets
 .PHONY: quickbyte quickopt                            # Partial / quick developer targets
-.PHONY: test-suite refman-html apidoc release         # Accesory targets
+.PHONY: test-suite refman-html stdlib-html apidoc release # Accesory targets
```
Probably to split this line in two, including one for Documentation targets.

ejgallego replied:
Indeed.

Zimmi48 commented on doc/stdlib/dune:
```diff
@@ -0,0 +1,20 @@
+; This is an ad-hoc rule to easy the migration, it should be handled
```
```suggestion
; This is an ad-hoc rule to ease the migration, it should be handled
```

ejgallego replied:
Fixed thanks.

ejgallego commented:
Will wait for the deploy build to finish before the rebase.

Zimmi48 commented:
There is also a problem with the artifacts (which should in principle make the deploy job fail): `_build/default/doc/stdlib/html/` is not exported. See https://gitlab.com/coq/coq/-/jobs/167795044/artifacts/browse/_build/default/doc/.

ejgallego commented:
Well-spotted @Zimmi48 ahhh; indeed @SkySkimmer was right, I've added an extra CI job which should help consistency.

Fixed version uploaded, deploy job needs relaunch.

SkySkimmer commented on .gitlab-ci.yml:
```diff
@@ -346,6 +346,15 @@ doc:refman:dune:
+    DUNE_TARGET: stdlib-html
+  artifacts:
+    <<: *dune-ci-template-artifacts
+      - _build/default/doc/stdlib/html
```
Missing a `paths` line

ejgallego replied:
Ups, thanks !

vbgl commented:
Pipeline launched for 7ba806bb.

vbgl commented:
The deploy jobs went through (also for the last update). There are some issues with the documentation of the stdlib. E.g., at https://coq.github.io/doc/dune+coqdoc/stdlib/, in the “navigation” part, there is a dead link to `genindex.html`.

ejgallego commented:
> The deploy jobs went through (also for the last update). There are some issues with the documentation of the stdlib. E.g., at https://coq.github.io/doc/dune+coqdoc/stdlib/, in the “navigation” part, there is a dead link to `genindex.html`.

It seems to work well here @vbgl , are you sure you were not looking to an older build [which had this problem]

So far, the docs looks pretty good to me.

vbgl commented:
My bad. It looks good indeed.

ejgallego commented:
Thanks to all for the help!
